README.md
[![Build Status](https://circleci.com/gh/StackStorm-Exchange/stackstorm-activedirectory.svg?style=shield&circle-token=:circle-token)](https://circleci.com/gh/StackStorm-Exchange/stackstorm-activedirectory) [![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)


# Microsoft Active Directory Integration Pack

# <a name="Introduction"></a> Introduction
This pack provides an integration between StackStorm and Microsoft Active Directory.
It is designed to mimic the Active Directory Cmdlets for PowerShell:

- [Server 2012 Docs](https://technet.microsoft.com/en-us/library/ee617195.aspx)
- [Server 2016 Docs](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/activedirectory)

This pack works by executing Active Directory PowerShell commands on a remote
windows hosts.


# <a name="QuickStart"></a> Quick Start


**Steps**

1. Install the pack

    ``` shell
    st2 pack install activedirectory
    ```

2. Configure WinRM on a remote Windows host by running the [setup PowerShell
   script](https://github.com/ansible/ansible/blob/devel/examples/scripts/ConfigureRemotingForAnsible.ps1)

   ``` PowerShell
   Invoke-WebRequest https://github.com/ansible/ansible/blob/devel/examples/scripts/ConfigureRemotingForAnsible.ps1 -OutFile "ConfigureRemotingForAnsible.ps1"
   .\ConfigureRemotingForAnsible.ps1
   ```

3. Install Remote Server Administration Tools (RSAT):

    ``` shell
    st2 run activedirectory.install_rsat_ad_powershell hostname='remotehost.domain.com' username='Administrator' password='xxx'
    ```

4. Execute an action (example: check if HOSTTOGET is a member of AD)

    ``` shell
    st2 run activedirectory.get_ad_computer hostname='remotehost.domain.com' username='Administrator' password='xxx' args='HOSTTOGET'
    ```



# <a name="Prerequisites"></a> Prerequisites
This pack works by executing PowerShell commands on a remote Windows host that
has the following setup:

1. WinRM needs to be configured
   Execute the following script on all hosts that this pack will be running
   commands on:
   https://github.com/ansible/ansible/blob/devel/examples/scripts/ConfigureRemotingForAnsible.ps1

2. Install Remote Server Administration Tools (RSAT) tools for ActiveDirectory
   Install the "Active Directory Domain Services (AD DS) Tools and Active
   Directory Lightweight Directory Services (AD LDS) Tools" component of
   RSAT on all hosts that this pack will be running commands on.
   **Note** : This only works on Windows Server OSes.

## Manual Install
``` PowerShell
Import-Module Servermanager
Install-WindowsFeature -Name RSAT-AD-PowerShell
```

## Install using st2 action (from this pack)
``` shell
st2 run activedirectory.install_rsat_ad_powershell hostname='remotehost.domain.com' username='xxx' password='xxx'
```

# <a name="Installation"></a> Installation
This pack is hosted on the [stackstork-exchange](https://exchange.stackstorm.org/#), you can install it like so:

``` shell
st2 pack install activedirectory
```

# <a name="Configuration"></a> Configuration
You will need to specify Active Directory credentials that will be
used to connect to the remote Windows hosts in the
`/opt/stackstorm/config/activedirectory.yaml` file. You can specify multiple
sets of credentials using nested values.

**Note** : `st2 pack config` doesn't handle nested schemas very well (known bug)
    so it's best to create the configuration file yourself and copy it into
    `/opt/stackstorm/configs/activedirectory.yaml` then run `st2ctl reload --register-configs`


## <a name="Schema"></a> Schema

``` yaml
poll_interval: <Time between polling for the sensor: default =  120>
hostname: <Hostname to use in sensor>
groups: <Groups to check for change in membership>
sensor_credential_name: <Which credential name to use for authorizing the sensor>
output: <Output format of cmdlets exected by this pack (json, raw) (global): default = 'json'>
port: <default port number to use for WinRM connections: default = '5986'>
transport: <default transport to use for WinRM connections: default = 'ntlm'>

activedirectory:
  <credential-name-1>:
    username: <username@domain.tld (preferred) or domain\username>
    password: <password for username>
  <credential-name-2>:
    username: <username@domain.tld (preferred) or domain\username>
    password: <password for username>
    port: <port number override to use for WinRM connections: default = '5986'>
    transport: <transport override to use for WinRM connections: default = 'ntlm'>
  ...
```

## <a name="SchemaExample"></a> Schema Example
``` yaml
poll_interval: 120
hostname: 4.4.4.4
groups:
    - "Domain Admins"
sensor_credential_name: dev
port: 5986
transport: ntlm
  
activedirectory:
  dev:
    username: username@dev.domain.tld
    password: xxx
  test:
    username: username@test.domain.tld
    password: xxx
    port: 5522
  prod:
    username: produser@domain.tld
    password: xxx
    port: 6611
    transport: kerberos
  prod-svc:
    username: prodsvc@domain.tld
    password: xxx
    port: 1234
    transport: basic
```

**Note** : All actions allow you to specify a set of credentials as inputs
           to the action instead of requiring a configuration. See the [Actions](#Actions)
           section for more information.

# <a name="Actions"></a> Actions

Actions in this pack are based off of the PowerShell cmdlets in ActiveDirectory module [link](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/activedirectory).

The actions from these cmdlets are auto-generated based on a the file `etc/cmdlets.txt`
using the generation script `etc/cmdlets_generate.py`

Below is a list of the currently implemented actions based on cmdlets.

| Action | PowerShell Cmdlet | Description |
|--------|-------------------|-------------|
| add_ad_computer_service_account | [Add-ADComputerServiceAccount](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/add-adcomputerserviceaccount)  | Adds one or more service accounts to an Active Directory computer. |
| add_ad_domain_controller_password_replication_policy | [Add-ADDomainControllerPasswordReplicationPolicy](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/add-addomaincontrollerpasswordreplicationpolicy)  | Adds users, computers, and groups to the Allowed List or the Denied List of the read-only domain controller (RODC) Password Replication Policy (PRP). |
| add_ad_fine_grained_password_policy_subject | [Add-ADFineGrainedPasswordPolicySubject](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/add-adfinegrainedpasswordpolicysubject)  | Applies a fine-grained password policy to one more users and groups. |
| add_ad_group_member | [Add-ADGroupMember](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/add-adgroupmember)  | Adds one or more members to an Active Directory group. |
| add_ad_principal_group_membership | [Add-ADPrincipalGroupMembership](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/add-adprincipalgroupmembership)  | Adds a member to one or more Active Directory groups. |
| clear_ad_account_expiration | [Clear-ADAccountExpiration](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/clear-adaccountexpiration)  | Clears the expiration date for an Active Directory account. |
| disable_ad_account | [Disable-ADAccount](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/disable-adaccount)  | Disables an Active Directory account. |
| disable_ad_optional_feature | [Disable-ADOptionalFeature](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/disable-adoptionalfeature)  | Disables an Active Directory optional feature. |
| enable_ad_account | [Enable-ADAccount](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/enable-adaccount)  | Enables an Active Directory account. |
| enable_ad_optional_feature | [Enable-ADOptionalFeature](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/enable-adoptionalfeature)  | Enables an Active Directory optional feature. |
| get_ad_account_authorization_group | [Get-ADAccountAuthorizationGroup](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adaccountauthorizationgroup)  | Gets the Active Directory security groups that contain an account. |
| get_ad_account_resultant_password_replication_policy | [Get-ADAccountResultantPasswordReplicationPolicy](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adaccountresultantpasswordreplicationpolicy)  | Gets the resultant password replication policy for an Active Directory account. |
| get_ad_computer | [Get-ADComputer](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adcomputer)  | Gets one or more Active Directory computers. |
| get_ad_computer_service_account | [Get-ADComputerServiceAccount](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adcomputerserviceaccount)  | Gets the service accounts that are hosted by an Active Directory computer. |
| get_ad_default_domain_password_policy | [Get-ADDefaultDomainPasswordPolicy](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-addefaultdomainpasswordpolicy)  | Gets the default password policy for an Active Directory domain. |
| get_ad_domain | [Get-ADDomain](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-addomain)  | Gets an Active Directory domain. |
| get_ad_domain_controller | [Get-ADDomainController](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-addomaincontroller)  | Gets one or more Active Directory domain controllers, based on discoverable services criteria, search parameters, or by providing a domain controller identifier, such as the NetBIOS name. |
| get_ad_domain_controller_password_replication_policy | [Get-ADDomainControllerPasswordReplicationPolicy](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-addomaincontrollerpasswordreplicationpolicy)  | Gets the members of the Allowed List or the Denied List of the RODC PRP. |
| get_ad_domain_controller_password_replication_policy_usage | [Get-ADDomainControllerPasswordReplicationPolicyUsage](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-addomaincontrollerpasswordreplicationpolicyusage)  | Gets the resultant password policy of the specified ADAccount on the specified RODC. |
| get_ad_fine_grained_password_policy | [Get-ADFineGrainedPasswordPolicy](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adfinegrainedpasswordpolicy)  | Gets one or more Active Directory fine-grained password policies. |
| get_ad_fine_grained_password_policy_subject | [Get-ADFineGrainedPasswordPolicySubject](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adfinegrainedpasswordpolicysubject)  | Gets the users and groups to which a fine-grained password policy is applied. |
| get_ad_forest | [Get-ADForest](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adforest)  | Gets an Active Directory forest. |
| get_ad_group | [Get-ADGroup](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adgroup)  | Gets one or more Active Directory groups. |
| get_ad_group_member | [Get-ADGroupMember](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adgroupmember)  | Gets the members of an Active Directory group. |
| get_ad_object | [Get-ADObject](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adobject)  | Gets one or more Active Directory objects. |
| get_ad_optional_feature | [Get-ADOptionalFeature](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adoptionalfeature)  | Gets one or more Active Directory optional features. |
| get_ad_organizational_unit | [Get-ADOrganizationalUnit](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adorganizationalunit)  | Gets one or more Active Directory OUs. |
| get_ad_principal_group_membership | [Get-ADPrincipalGroupMembership](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adprincipalgroupmembership)  | Gets the Active Directory groups that have a specified user, computer, or group. |
| get_ad_root_dse | [Get-ADRootDSE](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adrootdse)  | Gets the root of a domain controller information tree. |
| get_ad_service_account | [Get-ADServiceAccount](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-adserviceaccount)  | Gets one or more Active Directory service accounts. |
| get_ad_user | [Get-ADUser](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-aduser)  | Gets one or more Active Directory users. |
| get_ad_user_resultant_password_policy | [Get-ADUserResultantPasswordPolicy](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/get-aduserresultantpasswordpolicy)  | Gets the resultant password policy for a user. |
| install_ad_service_account | [Install-ADServiceAccount](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/install-adserviceaccount)  | Installs an Active Directory service account on a computer. |
| move_ad_directory_server | [Move-ADDirectoryServer](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/move-addirectoryserver)  | Moves a domain controller in AD DS to a new site. |
| move_ad_directory_server_operation_master_role | [Move-ADDirectoryServerOperationMasterRole](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/move-addirectoryserveroperationmasterrole)  | Moves operation master (also known as flexible single master operations or FSMO) roles to an Active Directory domain controller. |
| move_ad_object | [Move-ADObject](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/move-adobject)  | Moves an Active Directory object or a container of objects to a different container or domain. |
| new_ad_computer | [New-ADComputer](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/new-adcomputer)  | Creates a new Active Directory computer. |
| new_ad_fine_grained_password_policy | [New-ADFineGrainedPasswordPolicy](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/new-adfinegrainedpasswordpolicy)  | Creates a new Active Directory fine-grained password policy. |
| new_ad_group | [New-ADGroup](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/new-adgroup)  | Creates an Active Directory group. |
| new_ad_object | [New-ADObject](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/new-adobject)  | Creates an Active Directory object. |
| new_ad_organizational_unit | [New-ADOrganizationalUnit](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/new-adorganizationalunit)  | Creates a new Active Directory OU. |
| new_ad_service_account | [New-ADServiceAccount](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/new-adserviceaccount)  | Creates a new Active Directory service account. |
| new_ad_user | [New-ADUser](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/new-aduser)  | Creates a new Active Directory user. |
| remove_ad_computer | [Remove-ADComputer](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/remove-adcomputer)  | Removes an Active Directory computer. |
| remove_ad_computer_service_account | [Remove-ADComputerServiceAccount](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/remove-adcomputerserviceaccount)  | Removes one or more service accounts from a computer. |
| remove_ad_domain_controller_password_replication_policy | [Remove-ADDomainControllerPasswordReplicationPolicy](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/remove-addomaincontrollerpasswordreplicationpolicy)  | Removes users, computers, and groups from the Allowed List or the Denied List of the RODC PRP. |
| remove_ad_fine_grained_password_policy | [Remove-ADFineGrainedPasswordPolicy](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/remove-adfinegrainedpasswordpolicy)  | Removes an Active Directory fine-grained password policy. |
| remove_ad_fine_grained_password_policy_subject | [Remove-ADFineGrainedPasswordPolicySubject](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/remove-adfinegrainedpasswordpolicysubject)  | Removes one or more users from a fine-grained password policy. |
| remove_ad_group | [Remove-ADGroup](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/remove-adgroup)  | Removes an Active Directory group. |
| remove_ad_group_member | [Remove-ADGroupMember](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/remove-adgroupmember)  | Removes one or more members from an Active Directory group. |
| remove_ad_object | [Remove-ADObject](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/remove-adobject)  | Removes an Active Directory object. |
| remove_ad_organizational_unit | [Remove-ADOrganizationalUnit](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/remove-adorganizationalunit)  | Removes an Active Directory OU. |
| remove_ad_principal_group_membership | [Remove-ADPrincipalGroupMembership](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/remove-adprincipalgroupmembership)  | Removes a member from one or more Active Directory groups. |
| remove_ad_service_account | [Remove-ADServiceAccount](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/remove-adserviceaccount)  | Removes an Active Directory service account. |
| remove_ad_user | [Remove-ADUser](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/remove-aduser)  | Removes an Active Directory user. |
| rename_ad_object | [Rename-ADObject](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/rename-adobject)  | Changes the name of an Active Directory object. |
| reset_ad_service_account_password | [Reset-ADServiceAccountPassword](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/reset-adserviceaccountpassword)  | Resets the service account password for a computer. |
| restore_ad_object | [Restore-ADObject](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/restore-adobject)  | Restores an Active Directory object. |
| search_ad_account | [Search-ADAccount](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/search-adaccount)  | Gets Active Directory user, computer, and service accounts. |
| set_ad_account_control | [Set-ADAccountControl](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/set-adaccountcontrol)  | Modifies user account control (UAC) values for an Active Directory account. |
| set_ad_account_expiration | [Set-ADAccountExpiration](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/set-adaccountexpiration)  | Sets the expiration date for an Active Directory account. |
| set_ad_account_password | [Set-ADAccountPassword](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/set-adaccountpassword)  | Modifies the password of an Active Directory account. |
| set_ad_computer | [Set-ADComputer](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/set-adcomputer)  | Modifies an Active Directory computer. |
| set_ad_default_domain_password_policy | [Set-ADDefaultDomainPasswordPolicy](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/set-addefaultdomainpasswordpolicy)  | Modifies the default password policy for an Active Directory domain. |
| set_ad_domain | [Set-ADDomain](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/set-addomain)  | Modifies an Active Directory domain. |
| set_ad_domain_mode | [Set-ADDomainMode](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/set-addomainmode)  | Sets the domain functional level for an Active Directory domain. |
| set_ad_fine_grained_password_policy | [Set-ADFineGrainedPasswordPolicy](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/set-adfinegrainedpasswordpolicy)  | Modifies an Active Directory fine-grained password policy. |
| set_ad_forest | [Set-ADForest](https://technet.microsoft.com/en-us/itpro/powershell/windows/addsadministration/set-adforest)  | Modifies an Active Directory forest. |

CHANGES.md
# Change Log

## v1.0.1

- Ensure stdout, stderr are not bytes objects.

## v1.0.0

- Drop Python 2.7 support

## v0.1.4

- Spelling updates in the README.md.
- Removed the incubator installation instructions.

## v0.1.3

- Added automatic change to `http` when using port `5985`.

  Contributed by Nick Maludy (Encore Technologies)
  
- Added catching of JSON parsing errors, so a failure to parse data from JSON
  does not fail the action.
  
  Contributed by Nick Maludy (Encore Technologies)

## v0.1.2

- Fixed `$ProgressPreference=false` by setting it to `$ProgressPreference = 'SilentlyContinue'`

  Contributed by Nick Maludy (Encore Technologies)

## v0.1.1

- In new versions of Powershell if you load a module some text is output into 
  the stream. Pywinrm doesn't handle this gracefully and it leaks into the 
  stderr (https://github.com/diyan/pywinrm/issues/169). This version fixes
  the issue by inserting `$ProgressPreference=false` at the beginning of the
  powershell script that's being run.
  
  Contributed by Brad Bishop (Encore Technologies) #1

## v0.1.0

Initial Revision


actions/lib/action.py
from winrm_connection import WinRmConnection
from st2common.runners.base_action import Action
import json
import traceback

# Note:  in order for this to work you need to run the following script on the
# host
#  https://github.com/ansible/ansible/blob/devel/examples/scripts/ConfigureRemotingForAnsible.ps1

CREDENTIALS_ITEMS = ['username', 'password']
TRANSPORT_ITEMS = ['port', 'transport']


class BaseAction(Action):

    def __init__(self, config):
        """Creates a new BaseAction given a StackStorm config object (kwargs works too)
        :param config: StackStorm configuration object for the pack
        :returns: a new BaseAction
        """
        super(BaseAction, self).__init__(config)
        self.connection = None

    def resolve_creds(self, **kwargs):
        """Parses and resolves connection credentials (creds) from kwargs that
        were passed in as parameters to the action.
        :returns: A dict containing the 'connect' and 'cmdlet' credentials to
        use for this action execution
        """
        creds_config = self.create_creds_spec(**kwargs)
        creds = {}
        for key, value in creds_config.items():
            creds[key] = self.resolve_creds_spec(value)
        return creds

    def get_arg(self, key, **kwargs):
        """Attempts to retrieve an argument from kwargs with key.
        :param key: the key of the argument to retrieve from kwargs
        :returns: The value of key in kwargs, if it exists, otherwise None
        """
        if key in kwargs:
            return kwargs[key]
        else:
            return None

    def create_creds_spec(self, **kwargs):
        """Creates two credentials specification from all of the values in
        kwargs, one spec for 'connection', the other for 'cmdlet'.
        :returns: a dict of an initialized credentials spec from kwargs
        """
        return {'connect': {'credential_name': self.get_arg('credential_name', **kwargs),
                            'username': self.get_arg('username', **kwargs),
                            'password': self.get_arg('password', **kwargs),
                            'required': True},
                'cmdlet': {'credential_name': self.get_arg('cmdlet_credential_name', **kwargs),
                           'username': self.get_arg('cmdlet_username', **kwargs),
                           'password': self.get_arg('cmdlet_password', **kwargs),
                           'required': False}}

    def resolve_creds_spec(self, creds_spec):
        """Takes a credentials sepcification and attempts to lookup varaibles
        given the following order:

        1) If they are set in the incoming creds_spec, meaning they were given
        as an action parameter
        2) Lookup the variable from the 'credential_name' value
        :param creds_spec: dict of the credentials spec for this login.
        :returns: a credentials spec dict that has been resolved given the order
        """
        creds = {}

        # if the user specified the "credential_name" parameter
        # then grab the credentials from the pack's config.yaml
        credential_name = creds_spec['credential_name']
        config_creds = None
        if credential_name:
            config_creds = self.config['activedirectory'].get(credential_name)
            if not config_creds:
                raise KeyError("config.yaml missing credential: activedirectory:%s"
                               % credential_name)

        creds['credential_name'] = credential_name

        # Override the items in creds read in from the config given the
        # override parameters from the action itself
        # Example:
        #   'username' parameter on the action will override the username
        #   from the credential. This is useful for running the action
        #   standalone and/or from the commandline
        for item in CREDENTIALS_ITEMS:
            if item in creds_spec and creds_spec[item]:
                # use creds from cmdline first (override)
                creds[item] = creds_spec[item]
            elif config_creds and item in config_creds and config_creds[item]:
                # fallback to creds in config
                creds[item] = config_creds[item]

            # ensure that creds has all items (if this credential is required)
            if ('required' in creds_spec and creds_spec['required'] and item not in creds):
                if credential_name:
                    raise KeyError("config.yaml missing: activedirectory:%s:%s"
                                   % (credential_name, item))
                else:
                    raise KeyError("missing action parameter %s" % item)

        # copy in all transport items into the credentials spec
        for item in TRANSPORT_ITEMS:
            if config_creds and item in config_creds and config_creds[item]:
                creds[item] = config_creds[item]

        return creds

    @staticmethod
    def default_transport():
        """ Returns the default transport for this WinRM connection
        :returns: 'ntlm'
        """
        return 'ntlm'

    @staticmethod
    def default_port():
        """ Returns the default port for this WinRM connection
        :returns: '5986'
        """
        return 5986

    def resolve_transport(self, connect_creds, **kwargs):
        """ Resolves the transport and port to use for the connection
        based on the following priorities:
        1) if port/transport specified as action params, use this
        2) if port/transport specified as params on the credentials in config
        3) if port/transport specified at root level in the config
        4) else, use the default port/transport (5986/ntlm)
        :param connect_creds: the connection credentials read in from the
        config.
        :returns: dictionary with 'transport' and 'port' set the resolved
        values
        :rtype: dictionary
        """
        resolved_transport = None
        resolved_port = None

        if 'port' in kwargs and kwargs['port']:
            resolved_port = kwargs['port']
        elif 'port' in connect_creds and connect_creds['port']:
            resolved_port = connect_creds['port']
        elif 'port' in self.config and self.config['port']:
            resolved_port = self.config['port']
        else:
            resolved_port = BaseAction.default_port()

        if 'transport' in kwargs and kwargs['transport']:
            resolved_transport = kwargs['transport']
        elif 'transport' in connect_creds and connect_creds['transport']:
            resolved_transport = connect_creds['transport']
        elif 'transport' in self.config and self.config['transport']:
            resolved_transport = self.config['transport']
        else:
            resolved_transport = BaseAction.default_transport()

        return {'transport': resolved_transport, 'port': resolved_port}

    def connect(self, hostname, transport, creds):
        """Connects to hostname given the transport authentication method and
        creds for login. If the connection is successful it is cached locally
        as self.connection and reused for all commands run via run_ad_cmdlet().
        :param hostname: Hostname to connect to
        :param transport: Authentication method to use aka "transport" (WinRM terminology)
        :param creds: Credentials to use for the connection
        """
        if not self.connection:
            self.connection = WinRmConnection(hostname=hostname,
                                              port=transport['port'],
                                              transport=transport['transport'],
                                              username=creds['username'],
                                              password=creds['password'])

    def resolve_output_ps(self, **kwargs):
        """Generates powershell code for the output format selected by
        the action (if specified as a param, otherwise looked up in config).
        :returns: a string (potentially empty) containing powershell code
        that will format the output to the appropriate format.
        """
        output_ps = ""
        output = self.get_arg('output', **kwargs)
        from_config = False
        if not output:
            if 'output' in self.config:
                output = self.config['output']
                from_config = True
            else:
                raise LookupError("'output' parameter not found on action AND "
                                  "'output' option is missing in config!")

        # @note not using Write-Error here because it prints out the full
        # error object instead of just the message to stdout (this screws
        # up our parsing from JSON).
        # Instead we use a method found here to utilizes another function
        # to write just a string to stderr
        # https://stackoverflow.com/questions/4998173/how-do-i-write-to-standard-error-in-powershell/15669365#15669365
        if output == 'json':
            output_ps = ("Try\n"
                         "{{\n"
                         "  {0} | ConvertTo-Json\n"
                         "}}\n"
                         "Catch\n"
                         "{{\n"
                         "  $formatted_output = ConvertTo-Json -InputObject $_\n"
                         "  $host.ui.WriteErrorLine($formatted_output)\n"
                         "  exit 1\n"
                         "}}")
        elif output == 'raw':
            output_ps = "{0}"
        else:
            if from_config:
                raise LookupError("Unknown 'output' type [{0}] from config "
                                  "(valid = json, raw)".format(output))
            else:
                raise LookupError("Unknown 'output' type [{0}] from action parameter "
                                  "(valid = json, raw)".format(output))

        return output_ps

    def parse_output(self, output_str, **kwargs):
        parsed_output = {}
        if not output_str:
            return parsed_output

        output = self.get_arg('output', **kwargs)
        from_config = False
        if not output:
            if 'output' in self.config:
                output = self.config['output']
                from_config = True
            else:
                raise LookupError("'output' parameter not found on action AND "
                                  "'output' option is missing in config!")

        if output == 'json':
            try:
                parsed_output = json.loads(output_str)
            except ValueError:
                tb = traceback.format_exc()
                parsed_output = {'parse_error': tb}
        elif output == 'raw':
            parsed_output = {}
        else:
            if from_config:
                raise LookupError("Unknown 'output' type [{0}] from config "
                                  "(valid = json, raw)".format(output))
            else:
                raise LookupError("Unknown 'output' type [{0}] from action parameter "
                                  "(valid = json, raw)".format(output))

        return parsed_output

    def run_ad_cmdlet(self, cmdlet, **kwargs):
        """Runs an Active Directory cmdlet on a remote host.
        :param cmdlet: cmdlet to execut on remote host
        """
        creds = self.resolve_creds(**kwargs)
        tport = self.resolve_transport(creds['connect'], **kwargs)

        """Added $ProgressPreference = 'SilentlyContinue' as first line
        because if powershell attempts to load any modules
        the progress of the module loading is then forwarded
        into stderr and causes stackstorm to fail the call.
        There is an open pull request to pywinrm:
        https://github.com/diyan/pywinrm/issues/169
        """
        powershell = "$ProgressPreference = 'SilentlyContinue';\n"
        cmdlet_args = kwargs['args'] if 'args' in kwargs else ''
        output_ps = self.resolve_output_ps(**kwargs)

        if 'username' in creds['cmdlet'] and 'password' in creds['cmdlet']:
            powershell += ("$securepass = ConvertTo-SecureString \"{3}\" -AsPlainText -Force;\n"
                          "$admincreds = New-Object System.Management.Automation.PSCredential(\"{2}\", $securepass);\n"  # noqa
                          "{0} -Credential $admincreds {1}"
                          "").format(cmdlet,
                                     cmdlet_args,
                                     creds['cmdlet']['username'],
                                     creds['cmdlet']['password'])
        else:
            powershell += '{0} {1}'.format(cmdlet, cmdlet_args)

        # add output formatters to the powershell code
        powershell = output_ps.format(powershell)

        # connect to server
        self.connect(kwargs['hostname'], tport, creds['connect'])

        # run powershell command
        ps_result = self.connection.run_ps(powershell)

        # Ensure stdout and stderr is always a string
        stdout = ps_result.std_out
        stderr = ps_result.std_err
        if isinstance(stdout, bytes):
            stdout = stdout.decode("utf-8")
        if isinstance(stderr, bytes):
            stderr = stderr.decode("utf-8")

        result = {'stdout': stdout,
                  'stderr': stderr,
                  'exit_status': ps_result.status_code,
                  'stdout_dict': self.parse_output(stdout, **kwargs),
                  'stderr_dict': self.parse_output(stderr, **kwargs)}

        if result['exit_status'] == 0:
            return (True, result)
        else:
            return (False, result)
